### PR TITLE
Referral tests

### PR DESCRIPTION
**Pull request by SrLeet03**

### CODE REVIEW

**Review by stewartie4 — CHANGES_REQUESTED** [review]
some lint issues

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  | referral-tests  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | nft-cache  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  | referral-tests  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  | referral-tests  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | nft-cache  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  | referral-tests  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | nft-cache  |
| 0dns  | staging  |

**Comment by Kishan-Dhakan** [comment]
Passing run: https://github.com/0chain/0box/actions/runs/5113448332

File: internal/api/util/client/zbox_client.go [hunk 1]
```diff
@@ -1781,3 +1781,120 @@ func (c *ZboxClient) GetGraphBlobberTotalRewards(t *test.SystemTest, blobberId s
 
 	return &data, resp, err
 }
+
+func (c *ZboxClient) GetReferralCode(t *test.SystemTest, csrfToken, idToken, phoneNumber string) (model.ReferralCodeOfUser, *resty.Response, error) {
+	t.Logf("Checking if wallet exists for [%v] using 0box...", phoneNumber)
```
Comment (Kishan-Dhakan) [inline internal/api/util/client/zbox_client.go:1786]: Fix logging

File: internal/api/util/client/zbox_client.go [hunk 2]
```diff
@@ -1781,3 +1781,120 @@ func (c *ZboxClient) GetGraphBlobberTotalRewards(t *test.SystemTest, blobberId s
...
+			"X-CSRF-TOKEN":       csrfToken,
+		},
+		RequiredStatusCode: 200,
+	}, HttpGETMethod)
+
+	return ReferralCodeOfUser, resp, err
+}
+
+func (c *ZboxClient) GetReferralCount(t *test.SystemTest, csrfToken, idToken, phoneNumber string) (model.ReferralCountOfUser, *resty.Response, error) {
+	t.Logf("Checking if wallet exists for [%v] using 0box...", phoneNumber)
```
Comment (Kishan-Dhakan) [inline internal/api/util/client/zbox_client.go:1812]: Fix logging

File: internal/api/util/client/zbox_client.go [hunk 3]
```diff
@@ -1781,3 +1781,120 @@ func (c *ZboxClient) GetGraphBlobberTotalRewards(t *test.SystemTest, blobberId s
...
+			"X-CSRF-TOKEN":       csrfToken,
+		},
+		RequiredStatusCode: 200,
+	}, HttpGETMethod)
+
+	return ReferralCountOfUser, resp, err
+}
+
+func (c *ZboxClient) GetReferralRank(t *test.SystemTest, csrfToken, idToken, phoneNumber string) (model.ReferralRankOfUser, *resty.Response, error) {
+	t.Logf("Checking if wallet exists for [%v] using 0box...", phoneNumber)
```
Comment (Kishan-Dhakan) [inline internal/api/util/client/zbox_client.go:1838]: Fix logging

File: tests/api_tests/0box_referral_test.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+)
+
+func Test0BoxReferral(testSetup *testing.T) {
+	t := test.NewSystemTest(testSetup)
+	t.SetSmokeTests("Post referrals with correct CSRF should work properly")
+
+	firebaseToken := authenticateWithFirebase(t, zboxClient.DefaultPhoneNumber)
+	t.TestSetup("Autenticate with firebase", func() {
+		firebaseToken = authenticateWithFirebase(t, zboxClient.DefaultPhoneNumber)
+	})
```
Comment (Kishan-Dhakan) [inline tests/api_tests/0box_referral_test.go:17]: Does not need to be done twice 🤔

**Review by Kishan-Dhakan — CHANGES_REQUESTED** [review]

**Review by Kishan-Dhakan — APPROVED** [review]
LGTM

File: tests/api_tests/0box_referral_test.go (hunk 4 above)
Comment (Kishan-Dhakan) [reply tests/api_tests/0box_referral_test.go]: resolved

File: internal/api/util/client/zbox_client.go (hunk 3 above)
Comment (Kishan-Dhakan) [reply internal/api/util/client/zbox_client.go]: resolved

File: internal/api/util/client/zbox_client.go (hunk 2 above)
Comment (Kishan-Dhakan) [reply internal/api/util/client/zbox_client.go]: resolved

File: internal/api/util/client/zbox_client.go (hunk 1 above)
Comment (Kishan-Dhakan) [reply internal/api/util/client/zbox_client.go]: resolved

**Pull request merged**